force-app/main/adapter/in/sobjects/contact/Contacts.cls

```diff
@@ -389,81 +391,103 @@ public inherited sharing class Contacts extends fflib_SObjects {
+        Map<Address__c, Address__c> existingAddressesByAddress =
+                Addresses.getExistingAddresses(contactAddressesByContact.values());
 
-                // if found a match
-                if (addrDup != null) {
-                    // record the match in the contact
-                    con.Current_Address__c = addrDup.Id;
+        // maps & lists to keep track of the addresses we must insert or update
+        Map<Contact, Address__c> contactAddressesToInsertByContact = new Map<Contact, Address__c>();
 
-                    //reset the Primary Address Type only; resetting other fields on Contact will create a new Address
-                    con.npe01__Primary_Address_Type__c = con.npe01__Primary_Address_Type__c == null
-                            ? addrDup.Address_Type__c
-                            : con.npe01__Primary_Address_Type__c;
+        for (Contact contact : contacts) {
+            Address__c newAddressFromContact = contactAddressesByContact.get(contact);
+            Address__c existingAddressFromContact = existingAddressesByAddress.get(newAddressFromContact);
 
-                    // Prevent an address that was just inserted by the BeforeInsert trigger from being udpated
-                    // a second time by the AfterInsert trigger.
-                }
+            // if found a match
+            if (contactAddressHasAddressMatch(existingAddressFromContact)) {
+                updateContactAddressFromExistingAddress(contact, existingAddressFromContact);
 
-                // no match found, and its an override just for this contact
-                else if (con.is_Address_Override__c) {
-                    // put it on the list of addresss to create now
-                    mapConAddrInsertNow.put(con, addrNew);
-                }
+                // Prevent an address that was just inserted by the BeforeInsert trigger from being udpated
+                // a second time by the AfterInsert trigger.
             }
 
-            if (mapConAddrInsertNow.size() > 0) {
-                UTIL_DMLService.insertRecords(mapConAddrInsertNow.values());
-                // Since we're inserting new Addresses in a Before Insert/Update scenario, set the flag to
-                // prevent the After Insert/Update Contact logic from executing when it's eventually called.
-                TDTM_ProcessControl.setRecursionFlag(TDTM_ProcessControl.flag.ADDR_hasRunContactOnBefore, true);
-                for (Contact con : mapConAddrInsertNow.keySet()) {
-                    Address__c addr = mapConAddrInsertNow.get(con);
-                    con.Current_Address__c = addr.Id;
-                }
+            // no match found, and its an override just for this contact
+            else if (contact.is_Address_Override__c) {
+                // put it on the list of addresss to create now
+                newAddressFromContact.Undeliverable__c = contact.Undeliverable_Address__c;
```

Comment:
Can this be moved to the NPSP_Address constructor so that the undeliverable value is already set inside newAddressFromContact and doesn't need to be explicitly set here?